File: LeetCodeProblemsGo/LeetCode1022SumOfRootToLeafBinaryNumbers/LeetCode1022SumOfRootToLeafBinaryNumbers.go
package LeetCode1022SumOfRootToLeafBinaryNumbers

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func dfs(node *TreeNode, current int, sum *int) {
	if node == nil {
		return
	}
	current = current<<1 | node.Val
	if node.Left == nil && node.Right == nil {
		*sum += current
		return
	}
	dfs(node.Left, current, sum)
	dfs(node.Right, current, sum)
}
func sumRootToLeaf(root *TreeNode) int {
	sum := 0
	dfs(root, 0, &sum)
	return sum
}

File: LeetCodeProblemsCMake/Problems/LeetCode1022SumOfRootToLeafBinaryNumbers.cpp
//
// Created by Saeed Jannati on 2/24/2026 A.
//

#include "LeetCode1022SumOfRootToLeafBinaryNumbers.h"

#include "../Auxilary/TreeNode.h"

void dfs(TreeNode* node,int current, int& sum) {
    if (node==nullptr) {
        return;
    }
    current= (current<<1)|(node->val);
    if (node->left==nullptr && node->right==nullptr) {
        sum+=current;
        return;
    }
    dfs(node->left,current,sum);
    dfs(node->right,current,sum);
}
int LeetCode1022SumOfRootToLeafBinaryNumbers::sumRootToLeaf(TreeNode *root) {
    int sum=0;
    dfs(root,0,sum);
    return sum;
}
